src/stream_management/mod_stream_management.erl

```diff
@@ -120,63 +171,712 @@ stale_h_config_spec() ->
 %% hooks handlers
 %%
 
+-spec user_send_packet(mongoose_acc:t(), mongoose_c2s_hooks:hook_params(), gen_hook:extra()) ->
+    gen_hook:hook_fn_ret(mongoose_acc:t()).
+user_send_packet(Acc, #{c2s_data := StateData}, _Extra) ->
+    case {get_mod_state(StateData), is_not_sm_element(Acc)} of
+        {#sm_handler{counter_in = Counter} = SmHandler, true} ->
+            NewSmHandler = SmHandler#sm_handler{counter_in = incr_counter(Counter)},
+            {ok, mongoose_c2s_acc:to_acc(Acc, state_mod, {?MODULE, NewSmHandler})};
+        {_, _} ->
+            {ok, Acc}
+    end;
+user_send_packet(Acc, _Params, _Extra) ->
+    {ok, Acc}.
+
+-spec user_receive_packet(mongoose_acc:t(), mongoose_c2s_hooks:hook_params(), gen_hook:extra()) ->
+    gen_hook:hook_fn_ret(mongoose_acc:t()).
+user_receive_packet(Acc, #{c2s_data := StateData, c2s_state := C2SState} = Params, Extra) ->
+    Check1 = is_conflict_incomming_acc(Acc, StateData),
+    Check2 = is_conflict_receiver_sid(Acc, StateData),
+    Check3 = is_not_sm_element(Acc),
+    case {Check1, Check2, Check3} of
+        {true, _, _} -> %% A race condition detected: same jid, but different sids
+            C2SSid = mongoose_c2s:get_sid(StateData),
+            OriginSid = mongoose_acc:get(c2s, origin_sid, undefined, Acc),
+            ?LOG_WARNING(#{what => conflict_check_failed,
+                           text => <<"Drop Acc that is addressed to another connection "
+                                     "(origin SID check failed)">>,
+                           c2s_sid => C2SSid, origin_sid => OriginSid,
+                           acc => Acc, state_name => C2SState, c2s_state => StateData}),
+            {stop, Acc};
+        {_, true, _} ->
+            C2SSid = mongoose_c2s:get_sid(StateData),
+            ReceiverSID = mongoose_acc:get(c2s, receiver_sid, undefined, Acc),
+            ?LOG_WARNING(#{what => conflict_check_failed,
+                           text => <<"Drop Acc that is addressed to another connection "
+                                     "(receiver SID check failed)">>,
+                           c2s_sid => C2SSid, receiver_sid => ReceiverSID,
+                           acc => Acc, state_name => C2SState, c2s_state => StateData}),
+            {stop, Acc};
+        {_, _, false} ->
+            {ok, Acc};
```

Comment:
This might be unreachable, because packets are never SM elements, right?

Reply:
Aaaactually, yes, we never route SM elements, this is unneeded 👍🏽